evm/src/executor/inspector/cheatcodes/expect.rs

```diff
@@ -26,71 +29,152 @@ static DUMMY_CREATE_ADDRESS: Address =
+        .unwrap_or_else(|| format!("0x{}", hex::encode(data)))
+}
+
+fn actual_error_from_bytes(data: Bytes) -> Bytes {
+    if data.len() >= 4 && matches!(data[..4].try_into(), Ok(ERROR_PREFIX | REVERT_PREFIX)) {
+        if let Ok(bytes) = Bytes::decode(&data[4..]) {
+            return bytes
+        }
+    }
+
+    data
+}
+
 #[instrument(skip_all, fields(expected_revert, status, retdata = hex::encode(&retdata)))]
 pub fn handle_expect_revert(
     is_create: bool,
     expected_revert: Option<&Bytes>,
+    expected_revert_address: Option<H160>,
     status: InstructionResult,
+    reverts_by_address: BTreeMap<H160, HashSet<Bytes>>,
     retdata: Bytes,
 ) -> Result<(Option<Address>, Bytes)> {
     trace!("handle expect revert");
 
-    ensure!(!matches!(status, return_ok!()), "Call did not revert as expected");
+    if let Some(expected_revert_address) = expected_revert_address {
+        return handle_expect_revert_with_address(
+            is_create,
+            expected_revert,
+            expected_revert_address,
+            reverts_by_address,
+        )
+    }
 
-    macro_rules! success_return {
-        () => {
-            Ok(if is_create {
-                (Some(DUMMY_CREATE_ADDRESS), Bytes::new())
-            } else {
-                trace!("successfully handled expected revert");
-                (None, DUMMY_CALL_OUTPUT.to_vec().into())
-            })
-        };
+    handle_expect_revert_with_data(is_create, expected_revert, status, retdata)
+}
+
+/// Verifies that `expected_revert_address` reverted with `expected_reverted` anywhere in the call
+/// stack during EVM execution.
+fn handle_expect_revert_with_address(
+    is_create: bool,
+    expected_revert: Option<&Bytes>,
+    expected_revert_address: H160,
+    reverts_by_address: BTreeMap<H160, HashSet<Bytes>>,
+) -> Result<(Option<Address>, Bytes)> {
+    let address_reverts = reverts_by_address.get(&expected_revert_address);
+
+    if address_reverts.is_none() {
+        return Err(fmt_err!(
+            "The expected revert address {:#?} did not revert",
+            expected_revert_address,
+        ))
     }
 
+    #[allow(clippy::mutable_key_type)]
+    let address_reverts = address_reverts.unwrap();
+
     // If None, accept any revert
     let expected_revert = match expected_revert {
         Some(x) => x,
-        None => return success_return!(),
+        None => return success_return!(is_create),
     };
 
-    if !expected_revert.is_empty() && retdata.is_empty() {
-        bail!("Call reverted as expected, but without data");
+    let address_reverts = address_reverts
+        .iter()
+        .map(|revert_data| actual_error_from_bytes(revert_data.clone()).to_string())
+        .collect::<HashSet<_>>();
+
+    if address_reverts.contains(&expected_revert.to_string()) {
+        success_return!(is_create)
     }
+    // Empty revert data
+    else if address_reverts.contains("0x") {
```

Comment:
In the logic without the address the data is compared as bytes and the `.is_empty` is called. Here I use `0x` because the stringified version of an empty byte data is `0x` but if you think that there is a better way to achieve this I'd be welcome. 

Also to add to the context of why I did it this way if I use bytes as the datatype for the parsed hashset clippy complains that I'm using a type with interior mutability as a key to get a value from the hash set so between 1) use ` #[allow(clippy::mutable_key_type)]` and 2) convert `Bytes` to another type to get rid of the interior mutability I chose the second one.